x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -288,30 +290,31 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
+            updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, indexShard.routingEntry().index());
         } catch (Exception e) {
             throw new IndexShardRestoreFailedException(indexShard.shardId(), "failed to restore snapshot [" + snapshotId + "]", e);
         }
-
-        maybeUpdateMappings(client, remoteClient, leaderIndex, indexShard.indexSettings());
     }
 
     @Override
     public IndexShardSnapshotStatus getShardSnapshotStatus(SnapshotId snapshotId, Version version, IndexId indexId, ShardId leaderShardId) {
         throw new UnsupportedOperationException("Unsupported for repository of type: " + TYPE);
     }
 
-    private void maybeUpdateMappings(Client localClient, Client remoteClient, Index leaderIndex, IndexSettings followerIndexSettings) {
-        ClusterStateRequest clusterStateRequest = CcrRequests.metaDataRequest(leaderIndex.getName());
-        ClusterStateResponse clusterState = remoteClient.admin().cluster().state(clusterStateRequest)
-            .actionGet(ccrSettings.getRecoveryActionTimeout());
-        IndexMetaData leaderIndexMetadata = clusterState.getState().metaData().getIndexSafe(leaderIndex);
-        long leaderMappingVersion = leaderIndexMetadata.getMappingVersion();
-
-        if (leaderMappingVersion > followerIndexSettings.getIndexMetaData().getMappingVersion()) {
-            Index followerIndex = followerIndexSettings.getIndex();
-            MappingMetaData mappingMetaData = leaderIndexMetadata.mapping();
-            PutMappingRequest putMappingRequest = CcrRequests.putMappingRequest(followerIndex.getName(), mappingMetaData);
-            localClient.admin().indices().putMapping(putMappingRequest).actionGet(ccrSettings.getRecoveryActionTimeout());
+    private void updateMappings(Client leaderClient, Index leaderIndex, long leaderMappingVersion,
+                                Client followerClient, Index followerIndex) {
+        final PlainActionFuture<IndexMetaData> indexMetadataFuture = new PlainActionFuture<>();
+        final long startTimeInNanos = System.nanoTime();
+        final Supplier<TimeValue> timeout = () -> {
+            final long elapsedInNanos = System.nanoTime() - startTimeInNanos;
+            return TimeValue.timeValueNanos(ccrSettings.getRecoveryActionTimeout().nanos() - elapsedInNanos);
+        };
+        CcrRequests.getIndexMetadata(leaderClient, leaderIndex, leaderMappingVersion, 0L, timeout, indexMetadataFuture);
+        final IndexMetaData leaderIndexMetadata = indexMetadataFuture.actionGet(ccrSettings.getRecoveryActionTimeout());
+        final MappingMetaData mappingMetaData = leaderIndexMetadata.mapping();
+        if (mappingMetaData != null) {
+            final PutMappingRequest putMappingRequest = CcrRequests.putMappingRequest(followerIndex.getName(), mappingMetaData);
+            followerClient.admin().indices().putMapping(putMappingRequest).actionGet(ccrSettings.getRecoveryActionTimeout());
```

Comment:
This is more of a comment for @tbrooks8, but I think we need to explicitly increase the masterNodeTimeout on this request (by default it's 30 seconds), as a busy master might now possibly fail a CCR recovery.

Reply:
I will add a meta issue task.